### PR TITLE
Ensure that containers do not get stuck in stopping

### LINKED ISSUE #19629: podman stuck stopping
### Issue Description

When using the systemd service file generated by `podman generate systemd` server reboots often get the container stuck in an invalid state, "stopping" a container in this state cannot be started, stopped, or otherwise manipulated beyond removal.

It is possible to rescue the container if the server is rebooted an additional time (requiring 2 reboots to return to a normal state - the initial reboot putting the container in this state, plus one to get it out of this state. 

To combat this I removed the PIDFile directive from the systemd service file, as another issue (not seeing it off hand) mentioned that the pid file being removed prior to the container being killed properly caused some unexpected behavior.

This seems to have solved the issue (3 reboots in, so small sample size, requiring further testing). 

podman version
Client:       Podman Engine
Version:      4.6.0
API Version:  4.6.0
Go Version:   go1.20.6
Built:        Fri Jul 21 08:23:26 2023
OS/Arch:      linux/amd64

rpm -q podman
podman-4.6.0-1.fc38.x86_64

### Steps to reproduce the issue

Steps to reproduce the issue
1. create container and run as a daemon (run -d)
2. create service file using podman generate systemd
3. set service in place, enable, etc
4. reboot
5. observe podman container "stopping"
6. reboot to restore normal function
7. remove PIDFile directive
8. reboot
9. observe podman container is no longer "stopping"

### Describe the results you received

Container was left in an unusable state using generated systemd service file

### Describe the results you expected

podman generate systemd should provide a file that will run the service without additional modifications.

### podman info output

```yaml
podman info
host:
  arch: amd64
  buildahVersion: 1.31.0
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - hugetlb
  - pids
  - rdma
  - misc
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.7-2.fc38.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.7, commit: '
  cpuUtilization:
    idlePercent: 75.61
    systemPercent: 6.97
    userPercent: 17.41
  cpus: 24
  databaseBackend: boltdb
  distribution:
    distribution: fedora
    variant: server
    version: "38"
  eventLogger: journald
  freeLocks: 1997
  hostname: (hostname)
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 6.4.10-200.fc38.x86_64
  linkmode: dynamic
  logDriver: journald
  memFree: 19607564288
  memTotal: 135142043648
  networkBackend: cni
  networkBackendInfo:
    backend: cni
    dns:
      package: podman-plugins-4.6.0-1.fc38.x86_64
      path: /usr/libexec/cni/dnsname
      version: |-
        CNI dnsname plugin
        version: 1.3.1
        commit: unknown
    package: |-
      podman-plugins-4.6.0-1.fc38.x86_64
      containernetworking-plugins-1.3.0-2.fc38.x86_64
    path: /usr/libexec/cni
  ociRuntime:
    name: crun
    package: crun-1.8.6-1.fc38.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.8.6
      commit: 73f759f4a39769f60990e7d225f561b4f4f06bcf
      rundir: /run/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +LIBKRUN +WASM:wasmedge +YAJL
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: passt-0^20230625.g32660ce-1.fc38.x86_64
    version: |
      pasta 0^20230625.g32660ce-1.fc38.x86_64
      Copyright Red Hat
      GNU Affero GPL version 3 or later <https://www.gnu.org/licenses/agpl-3.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: true
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.0-12.fc38.x86_64
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.3
  swapFree: 8589930496
  swapTotal: 8589930496
  uptime: 2h 35m 2.00s (Approximately 0.08 days)
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 19
    paused: 0
    running: 17
    stopped: 2
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphRootAllocated: 107321753600
  graphRootUsed: 98293788672
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 139
  runRoot: /run/containers/storage
  transientStore: false
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 4.6.0
  Built: 1689942206
  BuiltTime: Fri Jul 21 08:23:26 2023
  GitCommit: ""
  GoVersion: go1.20.6
  Os: linux
  OsArch: linux/amd64
  Version: 4.6.0
```

### Podman in a container

No

### Privileged Or Rootless

Privileged

### Upstream Latest Release

No

### Additional environment details

Host is a bare metal box in my own lab. 

### Additional information

Until removal of the PIDFile directive this occurred every time a reboot was issued for kernel updates

### PR DESCRIPTION
**Pull request by mheon**
The scenario for inducing this is as follows:
1. Start a container with a long stop timeout and a PID1 that ignores SIGTERM
2. Use `podman stop` to stop that container
3. Simultaneously, in another terminal, kill -9 `pidof podman` (the container is now in ContainerStateStopping)
4. Now kill that container's Conmon with SIGKILL.
5. No commands are able to move the container from Stopping to Stopped now.

The cause is a logic bug in our exit-file handling logic. Conmon being dead without an exit file causes no change to the state. Add handling for this case that tries to clean up, including stopping the container if it still seems to be running.

Fixes #19629

#### Does this PR introduce a user-facing change?

```release-note
Fixed a bug where Systemd-managed containers could be stuck in the Stopping state, unable to be restarted, if systemd killed the unit before `podman stop` finished stopping the container ([#19629](https://github.com/containers/podman/issues/19629)).
```

### CODE REVIEW

**Comment by mheon** [comment]
Still needs a test

File: libpod/container_internal.go [hunk 1]
```diff
@@ -1442,6 +1452,53 @@ func (c *Container) waitForConmonToExitAndSave() error {
...
+				// continue.
+				// If we wanted to be really fancy here, we
+				// could open a pidfd on container PID1 before
+				// this to get the real exit code... But I'm not
+				// that dedicated.
+				all, _ := c.stopWithAll()
+				if err := c.ociRuntime.StopContainer(c, 0, all); err != nil {
+					logrus.Errorf("Error stopping container %s after Conmon exited prematurely: %w", c.ID(), err)
+				}
+			}
```
Comment (Luap99) [inline libpod/container_internal.go:1475]: I am not aware of any case were conmon can leak the container main process as it uses PDEATHSIG so I like to have it clear in the comment that this likely only relevant in the pid=host case were process are not killed as there is no pid namespace which does it for us.

File: libpod/container_internal.go [hunk 2]
```diff
@@ -1442,6 +1452,53 @@ func (c *Container) waitForConmonToExitAndSave() error {
...
+
+			c.newContainerExitedEvent(c.state.ExitCode)
+
+			if err := c.save(); err != nil {
+				logrus.Errorf("Error saving container %s state after Conmon exited prematurely: %w", c.ID(), err)
+			}
+
+			if err := c.runtime.state.AddContainerExitCode(c.ID(), c.state.ExitCode); err != nil {
+				logrus.Errorf("Error saving container %s exit code after Conmon exited prematurely: %w", c.ID(), err)
+			}
```
Comment (Luap99) [inline libpod/container_internal.go:1495]: we don't have a real exit code so should we actually store this -1 exit code?

**Review by Luap99 — COMMENTED** [review]
logic looks good

File: libpod/container_internal.go (hunk 2 above)
Comment (mheon) [reply libpod/container_internal.go:1495]: Needed for certain cases in `podman run` - otherwise we'd pop up an angry, duplicate error about not having a sensible error code.

File: libpod/container_internal.go (hunk 1 above)
Comment (mheon) [reply libpod/container_internal.go:1475]: I've heard of this happening even with PDEATHSIG set from support. Not sure of exact circumstances, but my trust in Conmon's ability to take the container down on being killed is not absolute.

File: libpod/container_internal.go (hunk 2 above)
Comment (Luap99) [reply libpod/container_internal.go:1495]: Mhh, yeah I guess it keeps it more consitent internally given many assumptions on having the exit code available even when it is sort of a lie.

File: test/system/030-run.bats [hunk 3]
```diff
@@ -1465,4 +1465,34 @@ search               | $IMAGE           |
     is "$output" "Error.*: $expect" "podman emits useful diagnostic when no entrypoint is set"
 }
 
+@test "podman run - stopping loop" {
```
Comment (mheon) [inline test/system/030-run.bats:1468]: @edsantiago PTAL

File: test/system/030-run.bats [hunk 4]
```diff
@@ -1465,4 +1465,34 @@ search               | $IMAGE           |
...
+    $(PODMAN) stop testctr &
+    stop_pid=$!
+
+    while :; do
+	sleep 0.5
+	run_podman logs testctr
+	if [[ "$output" =~ READY ]]; then
+	    break
+	fi
+    done
```
Comment (edsantiago) [inline test/system/030-run.bats:1485]: LGTM but since you have to repush anyway, please use `wait_for_ready testctr` instead of this loop

File: test/system/030-run.bats [hunk 5]
```diff
@@ -1465,4 +1465,34 @@ search               | $IMAGE           |
...
 
+@test "podman run - stopping loop" {
+    skip_if_remote "this doesn't work with with the REST service"
+
+    run_podman run -t -d --name testctr --stop-timeout 240 $IMAGE sh -c 'trap "echo READY" INT; while:; do sleep 0.1; done'
+
+    run_podman inspect --format '{{ .State.ConmonPid }}' testctr
+    conmon_pid=$output
+
+    $(PODMAN) stop testctr &
```
Comment (edsantiago) [inline test/system/030-run.bats:1476]: Curly braces, not parens

File: test/system/030-run.bats [hunk 6]
```diff
@@ -1465,4 +1465,34 @@ search               | $IMAGE           |
...
+    while :; do
+	sleep 0.5
+	run_podman logs testctr
+	if [[ "$output" =~ READY ]]; then
+	    break
+	fi
+    done
+
+    kill -9 $(stop_pid)
+    kill -9 $(conmon_pid)
```
Comment (edsantiago) [inline test/system/030-run.bats:1488]: here too

**Review by edsantiago — CHANGES_REQUESTED** [review]
OMG wait no

File: test/system/030-run.bats [hunk 7]
```diff
@@ -1465,4 +1465,28 @@ search               | $IMAGE           |
     is "$output" "Error.*: $expect" "podman emits useful diagnostic when no entrypoint is set"
 }
 
+@test "podman run - stopping loop" {
+    skip_if_remote "this doesn't work with with the REST service"
+
+    run_podman run -t -d --name testctr --stop-timeout 240 $IMAGE sh -c 'trap "echo READY" INT; while:; do sleep 0.1; done'
```
Comment (edsantiago) [inline test/system/030-run.bats:1471]: Need a space between `while` and the colon.

You can run this yourself via
```console
$ ./bats --rootless 030:"stopping loop"
```
(I include --rootless simply to avoid the sudo password thing, because there's no difference here anyway)

Comment (edsantiago) [reply test/system/030-run.bats:1471]: You also need to `s/INT/TERM/`, and possibly nuke the `-t`, but I *still* can't get this test to work.

Comment (mheon) [reply test/system/030-run.bats]: Added another wait_for_ready to make sure `sh -c` is good to go before we trigger the first stop - maybe that'll do it. Will check if it fixed things tomorrow.

Comment (edsantiago) [reply test/system/030-run.bats]: Oh it's much worse than that. I lied: you can't use `wait_for_ready` because that function "helpfully" checks container status and bails if != running. 

I'm working on something. Will let you know.

Comment (edsantiago) [reply test/system/030-run.bats]: This _passes_, but I suspect it's not quite exactly what you had hoped for. In particular, the `-t0` in the last `podman stop`. I can't make anything work without that.
```bash
@test "podman run - stopping loop" {
    skip_if_remote "this doesn't work with with the REST service"

    run_podman run -d --name testctr --stop-timeout 240 $IMAGE sh -c 'echo READY; sleep 999'
    cid="$output"
    wait_for_ready testctr

    run_podman inspect --format '{{ .State.ConmonPid }}' testctr
    conmon_pid=$output

    ${PODMAN} stop testctr &
    stop_pid=$!

    timeout=20
    while :;do
        sleep 0.5
        run_podman container inspect --format '{{.State.Status}}' testctr
        if [[ "$output" = "stopping" ]]; then
            break
        fi
        timeout=$((timeout - 1))
        if [[ $timeout == 0 ]]; then
            run_podman ps -a
            die "Timed out waiting for testctr to acknowledge signal"
        fi
    done

    kill -9 ${stop_pid}
    kill -9 ${conmon_pid}

    PODMAN_TIMEOUT=5 run_podman 125 stop -t0 testctr
    is "$output" "Error: container $cid conmon exited prematurely, exit code could not be retrieved: internal libpod error" "correct error on missing conmon"

    # This should be safe because stop is guaranteed to call cleanup?
    run_podman inspect --format "{{ .State.Status }}" testctr
    is "$output" "stopped" "container has successfully transitioned to exited state after stop"
}
```

**Comment by mheon** [comment]
Oof. The cleanup in `ContainerStop()` only happens if there was no error - so the error we're expecting prevents it, and prevents container cleanup. Actual code bug, we need to trigger a cleanup manually in this case otherwise it won't happen.

**Comment by mheon** [comment]
This is ready now

File: test/system/030-run.bats [hunk 8]
```diff
@@ -1465,4 +1465,44 @@ search               | $IMAGE           |
...
+    kill -9 ${conmon_pid}
+
+    # Unclear why `-t0` is required here, works locally without.
+    # But it shouldn't hurt and does make the test pass...
+    PODMAN_TIMEOUT=5 run_podman 125 stop -t0 testctr
+    is "$output" "Error: container .* conmon exited prematurely, exit code could not be retrieved: internal libpod error" "correct error on missing conmon"
+
+    # This should be safe because stop is guaranteed to call cleanup?
+    run_podman inspect --format "{{ .State.Status }}" testctr
+    is "$output" "exited" "container has successfully transitioned to exited state after stop"
```
[the remark this answers is not in the document]
Comment (mheon) [reply test/system/030-run.bats:1505]: Yes. We want `exited`. That indicates that cleanup has run properly, the container is unmounted, networking is torn down. `stopped` means we're still mounted, and we don't want to be stuck there.

**Review by edsantiago — APPROVED** [review]

**Comment by mheon** [comment]
@containers/podman-maintainers PTAL

File: test/system/030-run.bats [hunk 9]
```diff
@@ -1465,4 +1465,44 @@ search               | $IMAGE           |
...
+            die "Timed out waiting for testctr to acknowledge signal"
+        fi
+    done
+
+    kill -9 ${stop_pid}
+    kill -9 ${conmon_pid}
+
+    # Unclear why `-t0` is required here, works locally without.
+    # But it shouldn't hurt and does make the test pass...
+    PODMAN_TIMEOUT=5 run_podman 125 stop -t0 testctr
```
Comment (Luap99) [inline test/system/030-run.bats:1500]: @edsantiago Why PODMAN_TIMEOUT=5?

**Review by Luap99 — APPROVED** [review]
LGTM
But I must say the stop logic is really not nice to follow/understand. I would not be surprised if there are more corner cases.

The fact that we now have a second place where the event exited event is generated and the state is reset doesn't make things better as well.

File: test/system/030-run.bats (hunk 9 above)
Comment (Luap99) [reply test/system/030-run.bats:1500]: It is confusing to readers that ask why is this here.
